app/scripts/controllers/metametrics.test.js

```diff
@@ -604,6 +605,69 @@ describe('MetaMetricsController', function () {
     });
   });
 
+  describe('trackEvent', function () {
+    it('should call idempotent method createEventFragment, and finalizeEventFragment if actionId is provided', function () {
+      this.timeout(7000);
+      const metaMetricsController = getMetaMetricsController({
+        participateInMetaMetrics: false,
+      });
+      const spyTrackEvent = sinon.spy();
+      metaMetricsController._trackEvent = spyTrackEvent;
```

Comment:
Nit: Spying on internal methods is not advisable. This makes the tests more likely to break in the future even for non-functional changes such as internal refactors, and it does not test the true outcome we want (that the event is sent successfully).

Our test suite should be a safety net, making it safe for us to refactor and change things, giving us confidence that the outcomes we want are preserved. Tests that act as obstacles to future changes, and that don't test outcomes, are counter to this goal. They can still test internal logic, which is good in that it increases confidence in the code as it exists today. But they no longer effectively improve our velocity and enable future changes.

We can make these tests outcome-based and avoid using internal functions by relying on `nock` instead. We can setup interceptors for expected network calls with `nock`, and verify that they have been used.

Reply:
+1
I have tried to improve the tests, with metrics events it is bit harder to write outcome based events because of asynchronous nature of most functions.